plugin/restore/mysql/mysql.go

```diff
@@ -40,9 +44,76 @@ func New(driver *mysql.Driver, instance *mysqlbinlog.Instance) *Restore {
 	}
 }
 
-// RestoreBinlog restores the database using incremental backup in time range of [config.Start, config.End).
-func (r *Restore) RestoreBinlog(ctx context.Context, config mysql.BinlogInfo) error {
-	return fmt.Errorf("Unimplemented")
+// ReplayBinlog replays the binlog about `originDatabase` from `startInfo.Position` to `stopTs`.
+func (r *Restore) ReplayBinlog(ctx context.Context, originDatabase, pitrDatabase, binlogDir string, startInfo mysql.BinlogInfo, stopTs int64) error {
+	db, err := r.driver.GetDbConnection(ctx, "")
+	if err != nil {
+		return fmt.Errorf("cannot get database connection, error: %w", err)
+	}
+
+	binlogNamePrefix, err := r.getBinlogNamePrefix(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot get the prefix of binlog name, error: %w", err)
+	}
+
+	if !strings.HasPrefix(startInfo.FileName, binlogNamePrefix) {
+		return fmt.Errorf("start binlog name must has the prefix: %s, but get: %s", binlogNamePrefix, startInfo.FileName)
+	}
+
+	startBinlogSeq, err := strconv.ParseInt(strings.TrimPrefix(startInfo.FileName, binlogNamePrefix), 10, 0)
+	if err != nil {
+		return fmt.Errorf("cannot parse the start binlog name [%s], error: %w", startInfo.FileName, err)
+	}
+
+	binlogFilesLocal, err := ioutil.ReadDir(binlogDir)
+	if err != nil {
+		return fmt.Errorf("cannot read directory %s, error %w", binlogDir, err)
+	}
+
+	var needReplay []string
+	for _, f := range binlogFilesLocal {
+		if f.IsDir() || !strings.HasPrefix(f.Name(), binlogNamePrefix) {
+			continue
+		}
+		// for mysql binlog, after the serial number reaches 999999, the next serial number will not return to 000000, but 1000000,
+		// so we cannot directly use string to compare lexicographical order.
+		binlogSeq, err := strconv.ParseInt(strings.TrimPrefix(f.Name(), binlogNamePrefix), 10, 0)
+		if err != nil {
+			return fmt.Errorf("cannot parse the binlog name [%s], error: %w", f.Name(), err)
+		}
+		if binlogSeq >= startBinlogSeq {
+			needReplay = append(needReplay, f.Name())
+		}
+	}
+	sort.Strings(needReplay)
+
+	stopTime := time.Unix(stopTs, 0)
+	args := []string{
+		fmt.Sprintf(`--rewrite-db="%s->%s"`, originDatabase, pitrDatabase),
+		fmt.Sprintf("--database=%s", pitrDatabase),
+		"--disable-log-bin",
+		fmt.Sprintf("--start-position=%d", startInfo.Position),
+		fmt.Sprintf(`--stop-datetime="%d-%d-%d %d:%d:%d"`, stopTime.Year(), stopTime.Month(), stopTime.Day(), stopTime.Hour(), stopTime.Minute(), stopTime.Second()),
+	}
+
+	for _, name := range needReplay {
+		args = append(args, filepath.Join(binlogDir, name))
+	}
+
+	var stdout bytes.Buffer
+	cmd := exec.Command(r.mysqlbinlog.GetPath(), args...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("cannot run %s, error: %w", cmd.String(), err)
+	}
+
+	stmt := stdout.String()
+	if _, err := db.ExecContext(ctx, stmt); err != nil {
```

Comment:
https://dev.mysql.com/doc/refman/8.0/en/point-in-time-recovery-binlog.html
> $> mysqlbinlog binlog.000001 binlog.000002 | mysql -u root -p

We should just pipe the output to MySQL client program and let it handles the text stream.

Reply:
Another issue of the current implementation is that the result stmt will bloat the RAM off.